modules/core/src/main/java/org/apache/ignite/internal/processors/cache/mvcc/PreviousQueries.java

```diff
@@ -0,0 +1,186 @@
+     */
+    synchronized void init(Collection<ClusterNode> nodes, Predicate<UUID> alivePredicate) {
+        assert !init && !done;
+
+        nodes.stream().map(ClusterNode::id).forEach(uuid -> active.putIfAbsent(uuid, new Node()));
+
+        active.entrySet().removeIf(e -> !alivePredicate.test(e.getKey()) || e.getValue().done());
+
+        if (active.isEmpty())
+            done = true;
+
+        init = true;
+    }
+
+    /**
+     * @param nodeId Node ID.
+     */
+    void onNodeFailed(@NotNull UUID nodeId) {
+        if (done())
+            return;
+
+        synchronized (this) {
+            if (init)
+                removeAndCheckDone(nodeId);
+        }
+    }
+
+    /**
+     * @param nodeId Node ID.
+     * @param qryId Query tracker Id.
+     */
+    void onQueryDone(@NotNull UUID nodeId, long qryId) {
+        if (!done())
+            onQueryDone0(nodeId, qryId);
+    }
+
+    /**
+     * @param nodeId Node ID.
+     * @param queryIds Query tracker Ids.
+     */
+    void addActiveQueries(@NotNull UUID nodeId, @NotNull GridLongList queryIds) {
+        if (!done())
+            addActiveQueries0(nodeId, queryIds);
+    }
+
+    /**
+     * @return {@code True} if all queries mapped on previous coordinator are done.
+     */
+    boolean done() {
+        return done;
+    }
+
+    /** */
+    private synchronized void onQueryDone0(@NotNull UUID nodeId, long qryId) {
+        assert qryId > 0;
+
+        Node node = active.get(nodeId);
+
+        if (node == null && !init)
+            active.put(nodeId, node = new Node());
+
+        if (node != null) {
+            Set<Long> cntrs = node.cntrs;
+
+            boolean wasNull = cntrs == null;
+
+            if (cntrs == null)
+                cntrs = node.cntrs = new HashSet<>();
+
+            if (wasNull || !cntrs.remove(qryId))
+                cntrs.add(-qryId);
+
+            if (init && node.done())
+                removeAndCheckDone(nodeId);
+        }
+    }
+
+    /** */
+    private synchronized void addActiveQueries0(@NotNull UUID nodeId, @NotNull GridLongList queryIds) {
+        Node node = active.get(nodeId);
+
+        if (node == null && !init)
+            active.put(nodeId, node = new Node());
+
+        if (node != null) {
+            Set<Long> cntrs = node.cntrs;
+
+            boolean wasNull = cntrs == null, hasQueries = false;
+
+            for (int i = 0; i < queryIds.size(); i++) {
+                long qryId = queryIds.get(i);
+
+                assert qryId > 0;
+
+                if (cntrs == null)
+                    cntrs = node.cntrs = new HashSet<>();
+
+                if (wasNull || !cntrs.remove(-qryId))
+                    hasQueries |= cntrs.add(qryId);
+            }
+
+            if (init && !hasQueries)
+                removeAndCheckDone(nodeId);
+            else
+                node.init = true;
+        }
+    }
+
+    /** */
+    private void removeAndCheckDone(@NotNull UUID nodeId) {
```

Comment:
'check' look misleading here as we actually 'mark' as done, but not make a check.
Would e.g. 'finish(UUID)' clearer?

Reply:
not sure, it is just a couple of doings which is being done in common. It's almost always difficult to come up with right name for such methods. There are two actions here: 1) removing the node from the active list, 2) checking whether the done flag can be set and setting it if it is so. I think the current name is more or less descriptive.